tools/PI/DevHome.PI/Controls/AddToolControl.xaml.cs

```diff
@@ -79,77 +128,295 @@ private void OkButton_Click(object sender, RoutedEventArgs e)
+                }
+            }
+
+            currentCount++;
+            var percentComplete = PackageProcessingEndIndex + ((k + 1) * ShortcutProcessingEndIndex / _itemCount);
+            progress.Report(percentComplete);
+
+            // Yield to make sure the UI thread can update the progress output.
+            await Task.Delay(1);
+        }
     }
 
-    private void UpdateSampleCommandline()
+    public void ProcessShortcut(string filePath)
     {
-        if (SampleCommandTextBox is null)
+        var appName = Path.GetFileNameWithoutExtension(filePath);
+
+        // Exclude Microsoft Virtual Desktop _shortcuts.
+        if (appName.Contains("Microsoft Virtual Desktop", StringComparison.OrdinalIgnoreCase))
         {
-            // The window is still initializing.
             return;
         }
 
-        var tool = GetCurrentToolDefinition();
-        if (tool is null)
+        var wshShell = new WshShell();
+        if (wshShell.CreateShortcut(filePath) is not IWshShortcut shortcut)
         {
+            _log.Error("Error getting shortcut");
             return;
         }
 
-        SampleCommandTextBox.Text = tool.CreateFullCommandLine(123, (Windows.Win32.Foundation.HWND)123);
+        // Proceed with using the shortcut object.
+        var targetPath = shortcut.TargetPath;
+
+        // Exclude _shortcuts that point to empty targets, filesystem folders, or WSA apps.
+        if (string.IsNullOrEmpty(targetPath)
+            || Directory.Exists(targetPath)
+            || targetPath.Contains("WsaClient.exe", StringComparison.OrdinalIgnoreCase))
+        {
+            return;
+        }
+
+        // Exclude *.chm, *.url, *.html, *.ico targets.
+        var extension = Path.GetExtension(targetPath);
+        if (extension.Equals(".chm", StringComparison.OrdinalIgnoreCase)
+            || extension.Equals(".url", StringComparison.OrdinalIgnoreCase)
+            || extension.Equals(".html", StringComparison.OrdinalIgnoreCase)
+            || extension.Equals(".ico", StringComparison.OrdinalIgnoreCase))
+        {
+            return;
+        }
+
+        // Add the app for this shortcut to the list.
+        _allApps.Add(new InstalledAppInfo
+        {
+            Name = appName,
+            ShortcutFilePath = filePath,
+            TargetPath = targetPath,
+        });
     }
 
-    private void CancelButton_Click(object sender, RoutedEventArgs e)
+    public void ProcessPackage(Package package)
     {
-        ClearValues();
+        var op = package.GetAppListEntriesAsync();
+        var task = op.AsTask();
+        task.Wait();
+        var entries = task.Result;
+
+        // We only get the icon for apps that have an AppListEntry,
+        // because the others are not likely to be activatable from the UI.
+        if (entries.Count > 0)
+        {
+            // We only get the icon for the first AppListEntry, ignoring MAPs.
+            // Note we use Package.Logo, not AppListEntry.DisplayInfo.GetLogo
```

Comment:
should we just add all of them?  should we include appexecutonaliases as well?

Reply:
I filed a separate item to deal with appexecutionaliases, because there's more work to handle potentially multiple aliases, plus args: https://github.com/microsoft/devhome/issues/3251. Filed another one for handling MAPS: https://github.com/microsoft/devhome/issues/3270

Reply:
sounds good